### PR TITLE
nova: Added placement api service

### PR DESCRIPTION
**Pull request by nicolasbock**
The nova-placement-api service is optional in Newton and mandatory in
Ocata. This patch enables it.

### CODE REVIEW

File: chef/cookbooks/nova/recipes/placement_api.rb [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+      unless Dir.exist?("/srv/www/nova-placement-api")
+
+    # The wsgi script is installed in /usr/bin. We copy that script to make
+    # its name consistent with what the crowbar_openstack_wsgi resource
+    # expects. We should think about doing that in the package instead.
+    FileUtils.copy_file("/usr/bin/nova-placement-api",
+                        "/srv/www/nova-placement-api/app.wsgi")
+    FileUtils.chmod(0o644, "/srv/www/nova-placement-api/app.wsgi")
+  end
+end
```
Comment (nicolasbock) [inline chef/cookbooks/nova/recipes/placement_api.rb]: This part is a little ugly. Any suggestions on how to beautify it are very welcome :smile:

**Comment by nicolasbock** [comment]
https://trello.com/c/XkbQk63v/2184-3-enable-new-nova-placement-api

File: chef/cookbooks/nova/attributes/default.rb [hunk 2]
```diff
@@ -142,6 +159,7 @@
 default[:nova][:ha][:ports][:objectstore] = 5553
 default[:nova][:ha][:ports][:novncproxy] = 5554
 default[:nova][:ha][:ports][:serialproxy] = 5556
+default[:nova][:ha][:ports][:placement_api] = 5557
```
Comment (toabctl) [inline chef/cookbooks/nova/attributes/default.rb]: this is used already: ```default[:nova][:ha][:ports][:ec2_api] = 5557 ```

File: chef/cookbooks/nova/attributes/default.rb [hunk 3]
```diff
@@ -21,6 +21,7 @@
 default[:nova][:debug] = false
 default[:nova][:max_header_line] = 16384
 default[:nova][:config_file] = "/etc/nova/nova.conf.d/100-nova.conf"
+default[:nova][:placement_config_file] = "/etc/nova/nova.conf.d/101-nova-placement.conf"
```
Comment (toabctl) [inline chef/cookbooks/nova/attributes/default.rb]: Why an extra config file? All other nova services use the 100-nova.conf file

File: chef/cookbooks/nova/attributes/default.rb [hunk 4]
```diff
@@ -115,8 +129,11 @@
 default[:nova][:novnc][:ssl][:certfile] = ""
 default[:nova][:novnc][:ssl][:keyfile] = ""
 
+default[:nova][:api][:bind_host] = "*"
```
Comment (toabctl) [inline chef/cookbooks/nova/attributes/default.rb]: how is this related to the placement api?

File: chef/cookbooks/nova/attributes/default.rb [hunk 5]
```diff
@@ -83,6 +84,19 @@
 default[:nova][:scheduler][:reserved_host_memory_mb] = 512
 
 #
+# Placement Settings
+#
+default[:nova][:placement_user] = "placement"
```
Comment (toabctl) [inline chef/cookbooks/nova/attributes/default.rb]: Can we use ```[:nova][:placement][:user]``` here? iirc could the placement api be splitted out of nova and that would make a transition a bit easier.

File: chef/cookbooks/nova/recipes/placement_api.rb [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+keystone_register "register placement service" do
+  protocol keystone_settings["protocol"]
+  insecure keystone_settings["insecure"]
+  host keystone_settings["internal_url_host"]
+  port keystone_settings["admin_port"]
+  auth register_auth_hash
+  service_name "placement"
+  service_type "compute"
```
Comment (toabctl) [inline chef/cookbooks/nova/recipes/placement_api.rb]: upstream uses as service_type also ```placement```

File: chef/cookbooks/nova/recipes/placement_api.rb (hunk 1 above)
Comment (toabctl) [inline chef/cookbooks/nova/recipes/placement_api.rb]: Could we solve this in the package maybe?

File: chef/data_bags/crowbar/migrate/nova/115_placement_api.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
+def upgrade(ta, td, a, d)
+  a["placement_db"] = ta["placement_db"]
+
+  if a["placement_db"]["password"].nil? || a["placement_db"]["password"].empty?
+    service = ServiceObject.new "fake-logger"
+    a["placement_db"]["password"] = service.random_password
```
Comment (toabctl) [inline chef/data_bags/crowbar/migrate/nova/115_placement_api.rb]: I think we usually do that in the rails app. See https://github.com/crowbar/crowbar-openstack/blob/master/crowbar_framework/app/models/nova_service.rb#L180

File: chef/cookbooks/nova/templates/default/nova-placement.conf.erb [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+project_domain_name = <%= @keystone_settings["admin_domain"] %>
+username = <%= @username %>
+user_domain_name = <%= @keystone_settings['admin_domain'] %>
+password = <%= @keystone_settings['service_password'] %>
+
+[placement_database]
+<% if @database_connection -%>
+connection = <%= @database_connection %>
+mysql_sql_mode =
+<% unless node[:nova][:api_db][:max_pool_size].nil? -%>
```
Comment (toabctl) [inline chef/cookbooks/nova/templates/default/nova-placement.conf.erb]: copy&paste error. It must be ```:placement_db``` instead of ```api_db```

File: chef/cookbooks/nova/templates/default/nova-placement.conf.erb [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+max_pool_size = <%= node[:nova][:api_db][:max_pool_size] %>
```
Comment (toabctl) [inline chef/cookbooks/nova/templates/default/nova-placement.conf.erb]: same here and in the following lines

File: chef/cookbooks/nova/recipes/placement_api.rb (hunk 1 above)
Comment (Itxaka) [inline chef/cookbooks/nova/recipes/placement_api.rb]:
```
directory '/srv/www/nova-placement-api' do
  owner 'root'
  group 'root'
  mode '0755'
  action :create
end

file '/srv/www/nova-placement-api/app.wsgi' do
  content IO.read('/usr/bin/nova-placement-api')
  mode '0644'
  action :create
end
```

File: chef/cookbooks/nova/recipes/placement_api.rb [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+  bind_port = node[:nova][:ha][:ports][:placement_api]
+else
+  bind_host = node[:nova][:api][:bind_host]
+  bind_port = node[:nova][:ports][:placement_api]
+end
+
+node.normal[:apache][:listen_ports_crowbar] ||= {}
+node.normal[:apache][:listen_ports_crowbar][:nova] = { plain: [bind_port] }
+
+ruby_block "setup nova-placement-api wsgi script" do
```
Comment (aplanas) [inline chef/cookbooks/nova/recipes/placement_api.rb]: I am not sure that you need that:

* you can fix it in the package
* or you can set the place where the alias in the lwrp that you use later

**Comment by toabctl** [comment]
Rebased and continued at https://github.com/crowbar/crowbar-openstack/pull/1057
